demos/exyielding/README.md

```diff
@@ -0,0 +1,25 @@
+# Exyielding
+
+In this demo we want to show that a thread can yield the utilization of the CPU directly to another thread.
+
+To illustrate the mechanism in it, we will explain:
+
+* `yield_to` (in `src/user/thread.c`)
+
+## `yield_to`
+
+When this function is called, the kernel will check if the thread we specified is available to yield to.
+
+If the thread can be yielded to, we will set `next_thread` to the address of this thread and call the `do_scheduler` function.
+
+Since we have decided `next_thread`, the scheduler will do housekeeping tasks and help us to setup the thread we had already chosen.
+
+If the thread we specified is unavailable to yield, `do_scheduler` will not be called and we will go back to the point we called `yield_to()`.
+
+## Walkthrough
+
+At the beginning of this demo, we create two named thread  (`first` and `second`). After the thread (`setup`) ends, the thread (`first`) will get the use of the CPU and yield to the thread `second` specified by its thread ID (`tid`).
+
+The thread `second` will use `yield_to()` to specify the thread to yield and use `yield()` to yield the use to other threads. After yielding twice to both threads (`first` and `last`), the two would have reached their end. The next time the thread `second` tries to do `yield_to()` or `yield()`, it should fail so the yielding will not occur.
+
+As a result, the thread `second` will exit normally, and the kernel will exit since there are no more threads to run.
```

Comment:
This is missing where `last` is created.

Granted, I can't remember why I added `last` to the demo. I think perhaps to check you could yield (aka yield next) to a thread with a higher tid as well as a lower tid. Which proves that we can walk `all_threads` around to the end and back to an earlier thread, or stop at a later thread. (I don't think the explanation needs to include this detail though)

Reply:
I am thinking that maybe we can create the thread `last` first then do yielding to the thread `first`. It will be more clear to show that we can yield to the higher or the lower `tid` as well.  Would it be better?